module/x/gravity/abci.go

```diff
@@ -100,12 +101,50 @@ func pruneSignerSetTxs(ctx sdk.Context, k keeper.Keeper) {
 		earliestToPrune := currentBlock - params.SignedSignerSetTxsWindow
 		for _, set := range k.GetSignerSetTxs(ctx) {
 			if set.Nonce < lastObserved.Nonce && set.Height < earliestToPrune {
+				k.DeleteEthereumSignatures(ctx, set.GetStoreIndex())
 				k.DeleteOutgoingTx(ctx, set.GetStoreIndex())
 			}
 		}
 	}
 }
 
+// pruneCompletedOutgoingTxs deletes all completed txs and their signatures whos block height is below the last slashed
+// height. This accounts for the corner case where a tx becomes a CompletedOutgoingTx right after its relevant block height
+// has been slashed for, since it's possible for a relayer to submit a tx right before its slashing height.
+func pruneCompletedOutgoingTxs(ctx sdk.Context, k keeper.Keeper) {
+	lastSlashed := k.GetLastSlashedOutgoingTxBlockHeight(ctx)
+	k.IterateCompletedOutgoingTxs(ctx, func(key []byte, cotx types.OutgoingTx) bool {
+		if cotx.GetCosmosHeight() <= lastSlashed {
+			k.DeleteCompletedOutgoingTx(ctx, cotx.GetStoreIndex())
+		}
+		return false
+	})
+
+}
+
+// pruneEventVoteRecords deletes all event vote records for events that are oustide of the event vote window
+func pruneEventVoteRecords(ctx sdk.Context, k keeper.Keeper) {
+	lastEthereumHeight := k.GetLastObservedEthereumBlockHeight(ctx).EthereumHeight
+	window := k.GetParams(ctx).EventVoteWindow
+	if lastEthereumHeight < window {
+		return
+	}
+	minEthereumHeight := lastEthereumHeight - window
+
+	k.IterateEthereumEventVoteRecords(ctx, func(key []byte, eventVoteRecord *types.EthereumEventVoteRecord) bool {
+		event, err := types.UnpackEvent(eventVoteRecord.Event)
+		if err != nil {
+			panic(err)
+		}
+
+		if event.GetEthereumHeight() < minEthereumHeight {
+			k.DeleteEthereumEventVoteRecord(ctx, event.GetEventNonce(), event.Hash())
```

Comment:
Hm. If the window is made too small or there is some widespread orchestrator issue, there is a small chance of unconfirmed but legitimate votes being pruned. On the other hand, requiring confirmation before pruning would be spammable unless we slash for fake event votes right?

Reply:
The thing though is that the bridge must proceed sequentially through Ethereum event nonces otherwise the bridge gets stuck. If for some reason an event goes unconfirmed past this window and the votes get automatically pruned, I'm not sure it could recover.

Reply:
Yeah you're right; once that Ethereum height goes by it can't be accepted by the chain. Will add that check

Reply:
This one isn't getting marked as outdated because the comment isn't on line 140 but I addressed it.